diff -ru libmikmod-3.1.12.orig/include/mikmod_internals.h libmikmod-3.1.12/include/mikmod_internals.h
--- libmikmod-3.1.12.orig/include/mikmod_internals.h	2009-10-05 03:36:00.000000000 -0400
+++ libmikmod-3.1.12/include/mikmod_internals.h	2012-01-14 19:45:21.000000000 -0500
@@ -51,7 +51,7 @@
 
 /* SLONGLONG: 64bit, signed */
 #if defined(__arch64__) || defined(__alpha) || defined(__x86_64) || defined(_LP64)
-typedef long		SLONGLONG;
+typedef long long	SLONGLONG;
 #define NATIVE_64BIT_INT
 #elif defined(__WATCOMC__)
 typedef __int64		SLONGLONG;
