diff -ru --no-dereference /tmp/SDL2-2.0.8/Android.mk ./Android.mk
--- /tmp/SDL2-2.0.8/Android.mk	2018-03-01 18:34:41.000000000 +0200
+++ ./Android.mk	2021-09-14 12:26:37.098124377 +0300
@@ -49,6 +49,7 @@
 	$(wildcard $(LOCAL_PATH)/src/test/*.c))
 
 LOCAL_CFLAGS += -DGL_GLEXT_PROTOTYPES
+LOCAL_CFLAGS += -include $(LOGUTILS)
 LOCAL_LDLIBS := -ldl -lGLESv1_CM -lGLESv2 -llog -landroid
 
 cmd-strip := 
@@ -77,7 +78,7 @@
 ###########################
 
 include $(CLEAR_VARS)
-
+LOCAL_CFLAGS += -include $(LOGUTILS)
 LOCAL_C_INCLUDES := $(LOCAL_PATH)/include
 
 LOCAL_MODULE := SDL2_main
diff -ru --no-dereference /tmp/SDL2-2.0.8/src/events/SDL_mouse.c ./src/events/SDL_mouse.c
--- /tmp/SDL2-2.0.8/src/events/SDL_mouse.c	2018-03-01 18:34:42.000000000 +0200
+++ ./src/events/SDL_mouse.c	2021-09-14 12:53:34.467540503 +0300
@@ -655,6 +655,10 @@
 void
 SDL_WarpMouseInWindow(SDL_Window * window, int x, int y)
 {
+#ifdef __LIMBO__
+    // LIMBO: we don't warp in limbo
+    return;
+#endif
     SDL_Mouse *mouse = SDL_GetMouse();
 
     if (window == NULL) {
