--- pcre_compile.c	2006-10-10 12:00:00.000000000 +0200
+++ pcre_compile.c	2006-10-10 12:00:00.000000000 +0200
@@ -589,7 +589,7 @@ while (bot < top)
 while (bot < top)
   {
   i = (bot + top) >> 1;
-  c = strcmp(name, _pcre_utt[i].name);
+  c = strcmp(name, &_pcre_ucp_names[_pcre_utt[i].offset]);
   if (c == 0)
     {
     *dptr = _pcre_utt[i].value;
--- pcre_internal.h	2006-10-10 12:00:00.000000000 +0200
+++ pcre_internal.h	2006-10-10 12:00:00.000000000 +0200
@@ -993,7 +993,7 @@ codes. */
 codes. */
 
 typedef struct {
-  const char *name;
+  pcre_uint16 offset;
   pcre_uint16 type;
   pcre_uint16 value;
 } ucp_type_table;
@@ -1011,6 +1011,7 @@ extern const uschar _pcre_utf8_table4[];
 
 extern const int    _pcre_utf8_table1_size;
 
+extern const char _pcre_ucp_names[];
 extern const ucp_type_table _pcre_utt[];
 extern const int _pcre_utt_size;
 
